### ISSUE TITLE
Complex control flow in DaphneLib (DAPHNE Python API)

### ISSUE THREAD

**Issue by pdamme** [daphne-eu/daphne#522]
**Motivation:** While DAPHNE has its own domain-specific language DaphneDSL, the system also offers a Python API (DaphneLib) to be accessible from Python scripts. DaphneLib employs lazy evaluation, i.e., operations are not executed immediately, but merely build up an internal DAG representation of the program. Only when the `compute()` function is called on the outcome of an expression, the entire DAG is transformed into a DaphneDSL script, which then goes through the entire DAPHNE compiler and runtime stack, including various optimizations of the program. This approach means that Python control flow structures cannot easily be used to create DaphneDSL control flow structures. For instance, the following DaphneLib/Python snippet:
```
dctx = DaphneContext()
X = dctx.fill(1.23, 5, 5)
for i in range(3):
    X = X + 1
print(X.compute())
```
does **not** result in the following DaphneDSL script:
```
X = fill(1.23, 5, 5);
for(i in 0:2)
    X = X + 1;
```
but in the following DaphneDSL script:
```
X = fill(1.23, 5, 5);
X = X + 1;
X = X + 1;
X = X + 1;
```
This is because Python control flow is evaluated eagerly.

**Task:** This project is to extend DaphneLib, DAPHNE’s Python API, by means to build various control structures in DaphneDSL. Typical constructs like if-then-else, for-loops, while-loops, and do-while-loops shall be supported. Furthermore, defining and calling user-defined functions (with lazily evaluated function bodies) should be possible.

**Hints:**
- You can potentially take inspiration from other Python libraries with lazy evaluation, such as TensorFlow.

**Comment by lachezar-n** [comment]
I was assigned and will work on this issue as part of the course AMLS this summer semester - 2023.

**Comment by pdamme** [comment]
Thanks for reaching out and for working on this issue. Feel free to approach me if you have any questions.

**Comment by lachezar-n** [comment]
I am trying to build the project locally to get first familiar with it, but I was not able to since this the following error occurred:
```
make[2]: execvp: Advanced_C+FORTRAN/sde_test_f08.F90: Permission denied
make[2]: *** [Makefile:151: sde_test_f08] Error 127
```
*I am building it on an Ubuntu 20.04 Server 

Did I miss some point in the documentation about the building?

**Comment by pdamme** [comment]
I haven't seen this error before. I assume it happened when running `./build.sh`. It is most likely a problem with one of the third-party dependencies. Could you provide the lines above the error (most importantly, the last line starting with `[DAPHNE]..`), please?

As an alternative, you could try running DAPHNE in a Docker container by following the instructions in [containers/Readme.md](https://github.com/daphne-eu/daphne/tree/main/containers). Within the container, you have a complete environment for building and running DAPHNE.

**Comment by lachezar-n** [comment]
Yes, it occurred when build with `./build.sh`. It seems that the problem is really with one of the third-party dependencies since the last directory entered is:
```
make[2]: Entering directory 
'/home/ubuntu/daphne/thirdparty/sources/papi-7.0.1/src/components/sde/tests' 
```

Here is the error log with the lines before:
```
g++ Counting_Set/MemoryLeak_CountingSet_Driver++.cpp -o Counting_Set_MemLeak_Test++ -I. -I../../.. 
-I../../../testlib -I../../../validation_tests -I/home/ubuntu/daphne/thirdparty/installed/include 
-I../../../sde_lib -I.. -g -std=c++11 ../../../testlib/libtestlib.a -lCounting_Set++ -lrt -ldl -Llib 
-L../../.. -L../../../libpfm4/lib -lpapi -lpfm -lsde
Advanced_C+FORTRAN/sde_test_f08.F90 -o sde_test_f08 -I. -I../../.. -I../../../testlib 
-I../../../validation_tests -I/home/ubuntu/daphne/thirdparty/installed/include -I../../../sde_lib -I.. 
-free -g ../../../testlib/libtestlib.a -lXandria -lGamum -lRecorder -lrt -ldl -Llib -L../../.. 
-L../../../libpfm4/lib -lpapi -lpfm -lsde
make[2]: execvp: Advanced_C+FORTRAN/sde_test_f08.F90: Permission denied
make[2]: *** [Makefile:151: sde_test_f08] Error 127
make[2]: Leaving directory '/home/ubuntu/daphne/thirdparty/sources/papi-7.0.1/src/components/sde/tests'
make[1]: *** [Makefile.inc:265: comp_tests] Error 2
make[1]: *** Waiting for unfinished jobs....
```
The log history if the terminal does not contain any line starting with `[DAPHNE]...` at all. (These are not the last log lines but the next are about finished the jobs.)

I want to avoid running Daphne in a Docker container since I already setup an AWS EC2 instance with Ubuntu 20.04 only for the project because my PC is Macbook with M-chip (arch64 architecture) and even docker building locally is not possible.

**Comment by corepointer** [comment]
Hi!
I just built papi again with build.sh and here the lines you mention read as follows:
```
g++ Counting_Set/MemoryLeak_CountingSet_Driver++.cpp -o Counting_Set_MemLeak_Test++ -I. -I../../.. -I../../../testlib -I../../../validation_tests -I/home/mdokter/code/daphne-eu/daphne/thirdparty/installed/include -I../../../sde_lib -I.. -g -std=c++11 ../../../testlib/libtestlib.a -lCounting_Set++ -lrt -ldl -Llib -L../../.. -L../../../libpfm4/lib -lpapi -lpfm -lsde
gfortran Advanced_C+FORTRAN/sde_test_f08.F90 -o sde_test_f08 -I. -I../../.. -I../../../testlib -I../../../validation_tests -I/home/mdokter/code/daphne-eu/daphne/thirdparty/installed/include -I../../../sde_lib -I.. -ffree-form -ffree-line-length-none -g ../../../testlib/libtestlib.a -lXandria -lGamum -lRecorder -lrt -ldl -Llib -L../../.. -L../../../libpfm4/lib -lpapi -lpfm -lsde
```
Please make sure that you have installed gfortran.
hth

**Comment by lachezar-n** [comment]
@corepointer Thanks for the advice, that helped to solve the mentioned issue!

However, I still encounter errors at building and was not able to figure out how to fix them. 
Here is the log at running `./build.sh`:
```
[DAPHNE]..No need to build Antlr4 again...............................[29.05.23 18:21:49]
[DAPHNE]..No need to download Catch2 again............................[29.05.23 18:21:49]
[DAPHNE]..No need to build OpenBlas again.............................[29.05.23 18:21:49]
[DAPHNE]..No need to download nlohmannjson again......................[29.05.23 18:21:49]
[DAPHNE]..No need to build Abseil again...............................[29.05.23 18:21:49]
[DAPHNE]..No need to build OpenMPI again..............................[29.05.23 18:21:49]
[DAPHNE]..No need to build GRPC again.................................[29.05.23 18:21:49]
[DAPHNE]..No need to build Arrow again................................[29.05.23 18:21:49]
[DAPHNE]..No need to build spdlog again...............................[29.05.23 18:21:49]
[DAPHNE]..No need to build eigen again................................[29.05.23 18:21:49]
[DAPHNE]..No need to build PAPI again.................................[29.05.23 18:21:49]
[DAPHNE]..No need to build MLIR/LLVM again............................[29.05.23 18:21:49]
[DAPHNE]..Build Daphne................................................[29.05.23 18:21:49]
CMake Error at CMakeLists.txt:29 (cmake_policy):
  Policy "CMP0116" is not known to this version of CMake.


-- Using MLIRConfig.cmake in: /home/ubuntu/daphne/thirdparty/installed/lib/cmake/mlir
-- Using LLVMConfig.cmake in: /home/ubuntu/daphne/thirdparty/installed/lib/cmake/llvm
-- Building with -fPIC
-- CMAKE_PREFIX_PATH:  /home/ubuntu/daphne/thirdparty/installed
-- Arrow version: 11.0.0
-- Found the Arrow shared library: /home/ubuntu/daphne/thirdparty/installed/lib/libarrow.so.1100.0.0
-- Found the Arrow import library: ARROW_IMPORT_LIB-NOTFOUND
-- Found the Arrow static library: /home/ubuntu/daphne/thirdparty/installed/lib/libarrow.a
-- Parquet version: 11.0.0
-- Found the Parquet shared library: /home/ubuntu/daphne/thirdparty/installed/lib/libparquet.so.1100.0.0
-- Found the Parquet import library: PARQUET_IMPORT_LIB-NOTFOUND
-- Found the Parquet static library: /home/ubuntu/daphne/thirdparty/installed/lib/libparquet.a
-- ANTLR4_JAR_LOCATION is /home/ubuntu/daphne/thirdparty/installed/share/antlr4/antlr-4.9.2-complete.jar
-- Antlr4 DaphneDSLGrammar - Building Common Include-, Source- and Tokenfiles, Visitor Include- and Sourcefiles
-- Antlr4 SQLGrammar - Building Common Include-, Source- and Tokenfiles, Visitor Include- and Sourcefiles
-- ANTLR4_SQL_DIR is /home/ubuntu/daphne/build/antlr4_generated_src/SQLGrammar
-- Configuring incomplete, errors occurred!
See also "/home/ubuntu/daphne/build/CMakeFiles/CMakeOutput.log".
See also "/home/ubuntu/daphne/build/CMakeFiles/CMakeError.log".
```

Does anyone of you have any idea what is the source of this issue?
(I have used `--clean` and `--cleanDeps` every time after failing the building).

**Comment by corepointer** [comment]
Did you check the version of CMake in use? The CMake that you can install with apt (version 3.16) is too old for DAPHNE. You'll have to use the snap version of it, which receives continuous updates.

**Comment by lachezar-n** [comment]
Yes, I used snap for installing Cmake and even forced installing version 3.17 (and not the current supported one 3.26):
```
ubuntu@ip-172-31-2-195:~$ cmake --version
cmake version 3.17.5
```

**Comment by pdamme** [comment]
That is, the latest version available via snap. I personally use cmake 3.26.3.

**Comment by corepointer** [comment]
woops - CMP0116 requires 3.20 as it seems. Interesting how this little "bug" could survive so long :D 
[x] fixed

**Comment by lachezar-n** [comment]
Cmake version 3.26 worked! Thank you!

**Comment by lachezar-n** [comment]
Hi, I am currently trying to understand how DaphneLib works overall so I can plan how to solve the task.
The `MultiReturn' operation node seems not to be supported in the current version since I can not find any initialisation or usage and `class OutputType(Enum)` does not have member `MULTI_RETURN`. Is that correct?

**Comment by pdamme** [comment]
Indeed, `MultiReturn` is not used at the moment, you can basically ignore it for now. However, in the future, DaphneLib will support built-in operations and user-defined functions that can return multiple results.

**Comment by lachezar-n** [comment]
I have another concern regarding the `class OutputType(Enum)` and the function `DaphneDSLScript._dfs_dag_nodes` on line 120: `if isinstance(input_node, DAGNode) and input_node._output_type == OutputType.LIST:` (https://github.com/daphne-eu/daphne/blob/dc553258e9694a5a14a4e862d03b592a245d5ac0/src/api/python/script_building/script.py#L120).
I assume that for the moment `isinstance(input_node, DAGNode)` is never evaluated to `True` with the current API implementation since then `input_node._output_type == OutputType.LIST` would lead to `AttributeError`. 
But is that then a bug, or do I miss the function if this `if` statement?

**Comment by pdamme** [comment]
As far as I see it, `input_node` will always be an instance of (a subclass of) `DAGNode`. However, I think it won't raise an `AttributeError`, because `DAGNode` does have an attrbute `_output_type`. (Or am I missing something here?)

Nevertheless, something else is slightly wrong here: we do not actually use the output type `OutputType.LIST` so far, since DAPHNE does not support lists (at least not yet). I assume this piece of code doesn't have any effect. For some background: We started implementing DAPHNE's Python API by reusing the code from Apache SystemDS and simplifying/modifying it for our purposes. The lines you mention are probably a left-over of this translation.

**Comment by lachezar-n** [comment]
I started by implementing construct for `for-loops` and my idea is to create the constructs directly from the `DaphneContext()` object and then to pass on OperationNode Object (Matrix/Scalar) that is handled inside the construct:
```python
...
dctx = DaphneContext()

X = dctx.fill(3.8, 5, 5)

lambda_body = lambda x, i: x + 2

def func_body(x, i):
    return x + i*i

loop = dctx.for_loop(X, lambda_body, start=1, end=10, 2)
result = loop.compute()
print(result)
```
Here `i` represents the iterator. I assume that only a single object for handling can be passed since the API can have at the moment only a single output value.
The generated `tmpdaphne.daphne` for `lambda_body` is:
```
V0=0+0;
V1=fill(3.8,5,5);
for(V0 in 1:10:2) {
    L0=V1+2;
    V1=L0;
}
saveDaphneLibResult(V1);
```
and for `func_body`:
```Daphne
V0=0+0;
V1=fill(3.8,5,5);
for(V0 in 1:10:2) {
    L0=V0*V0;
    L1=V1+L0;
    V1=L1;
}
saveDaphneLibResult(V1);
```
This, however,  required deleting the if statement mentioned earlier: 
https://github.com/daphne-eu/daphne/blob/dc553258e9694a5a14a4e862d03b592a245d5ac0/src/api/python/script_building/script.py#L120

Can you give me an opinion if this approach (passing the object for handling and having its manipulated value as an output) is acceptable and reasonable for you? If it is I would move forward with using the same pattern for `if-else` and `while-loop`.

**Comment by pdamme** [comment]
**This design looks good to me.** Indeed, TensorFlow's [`tf.while_loop()`](https://www.tensorflow.org/api_docs/python/tf/while_loop) works in a similar way. 

I recommend also adding a test case where a variable which is not modfied in the loop is used in the loop, such as
```python
X = dctx.fill(3.8, 5, 5)
Y = dctx.fill(3.8, 5, 5)
X = dctx.for_loop(X, lambda x, i: x + Y, start=1, end=10, 2)
res = X.compute()
```
This should generally be possible. Ideally the code calculating `Y` should end up before the for-loop in DaphneDSL. If that turns out to be too difficult, we could rely on the DAPHNE compiler to move loop-invariant code before the loop (that may not work in DaphneDSL yet, and addressing it is beyond the scope of this issue).

While-loops can surely be implemented in a similar way. For if-then-else, you could also take inspiration from TensorFlow's [`tf.cond()`](https://www.tensorflow.org/api_docs/python/tf/cond).

It's a bit unfortunate that DaphneLib does not support returning multiple values from a single operation yet, but maybe this is a good opportunity for adding such a feature :) . Being able to modify multiple variables in a loop would significantly improve the usefulness.

**Comment by lachezar-n** [comment]
I did indeed tried to follow the concept of Tensorflow and my plan for the if-else statements would also follow this. 

Regarding the Multi-Return functionality, I would like to implement it, but I am not sure enough if I will be able to do it until 04.07. if I am going to implement also the feature of defining and calling user-defined functions for lazy evaluation. 

Since the user-defined functions is a “should” requirement, would it be better to concentrate myself on for-loops, while-loops and if-else statements with multi-return options until the deadline? In this case, I would be happy to work on the user-defined functions feature after the deadline because of lack of time.

**Comment by pdamme** [comment]
Thanks for reporting this problem, I can confirm it. I have opened a new issue (#560) for it. We will fix it on the main branch at some point. If it is blocking you, please use the workaround of inserting the spaces in DaphneLib.

**Comment by lachezar-n** [comment]
Since I was not sure how to handle MULTI_RETURN and did not really understand if this is working in DaphnaDSL (at least in my tests it didn't), I came up with a workaround that still allows multiple input nodes for handling inside the different functions, but the output itself is a python tuple of Matrices (for the user-defined functions tuple of Operation nodes). The objects from the tuples can be then used in operations together or with other Operation nodes before computing. The tuples themselves do not allow direct computing of all of them together. The tuple consists of Matrices following the concept of TensorFlow using instances Tensors.
A brief description of the implemented functions is to be found here: https://github.com/lachezar-n/daphne/blob/522-DaphneLib-complex-control-flow/doc/DaphneLibControlFlow.md
Can you have a look over it? And what are the required steps before making the pull request?

*All the implementation is to be found in the same branch and simple examples used at developing are to be found here: https://github.com/lachezar-n/daphne/blob/f77aea6efaa6badd95a68cb495dc67d1222386f6/probe.py. Variables not modified by the loop/statement bodies are evaluated up front as requested before.

**Comment by lachezar-n** [comment]
Hey!
I was testing and improving some bugs in the last few days and a potential bug in DaphneDSL occurred. My idea was to make possible even nesting complex control functions. However, the following snippet generated from DaphneLib looks fine to me and follows the logic for the complex control flow that I implemented but it leads to an error from the Daphne compiler:
```
V0=fill(3.8,5,5);
V1=0;
for(V1 in 1:10) {
    L1_V0=sum(V0);
    L1_V1=L1_V0 < 10;
    if (L1_V1) {
        L2_V0=V0 - 1;
        V0=L2_V0;
    } else {
        L2_V0=V0 + 1;
        V0=L2_V0;
    }
    L1_V3=V0;
    V0=L1_V3;
}
V3=V0;
saveDaphneLibResult(V3);
```
The occurred error is:
```
loc("tmpdaphne.daphne":13:0): error: operand #0 does not dominate this use
"builtin.module"() ({
  "func.func"() ({
    %0 = "daphne.constant"() {value = 6 : index} : () -> index
    %1 = "daphne.constant"() {value = 1 : index} : () -> index
    %2 = "daphne.constant"() {value = 11 : index} : () -> index
    %3 = "daphne.constant"() {value = 1 : si64} : () -> si64
    %4 = "daphne.constant"() {value = 5 : index} : () -> index
    %5 = "daphne.constant"() {value = 3.800000e+00 : f64} : () -> f64
    %6 = "daphne.fill"(%5, %4, %4) : (f64, index, index) -> !daphne.Matrix<5x5xf64>
    "scf.for"(%1, %2, %1) ({
    ^bb0(%arg0: index):
      %7 = "daphne.cast"(%arg0) : (index) -> si64
      %8 = "daphne.ewMul"(%7, %3) : (si64, si64) -> si64
      %9 = "scf.for"(%1, %0, %1, %6) ({
      ^bb0(%arg1: index, %arg2: !daphne.Matrix<5x5xf64>):
        %10 = "daphne.cast"(%arg1) : (index) -> si64
        %11 = "daphne.ewMul"(%10, %3) : (si64, si64) -> si64
        %12 = "daphne.ewAdd"(%arg2, %3) : (!daphne.Matrix<5x5xf64>, si64) -> !daphne.Matrix<5x5xf64>
        "scf.yield"(%12) : (!daphne.Matrix<5x5xf64>) -> ()
      }) : (index, index, index, !daphne.Matrix<5x5xf64>) -> !daphne.Matrix<5x5xf64>
      "scf.yield"() : () -> ()
    }) : (index, index, index) -> ()
    "daphne.saveDaphneLibResult"(%9) : (!daphne.Matrix<5x5xf64>) -> ()
    "daphne.return"() : () -> ()
  }) {function_type = () -> (), sym_name = "main"} : () -> ()
}) : () -> ()
loc("tmpdaphne.daphne":0:0): error: module pass error
```
The same error occurred when I tried nesting different statements and loops.
Is that really a bug or do I just miss the source of the error in the generated code?


*I am writing pytests at the moment and I created several testing functions for nesting that are being skipped at the moment.

**Comment by pdamme** [comment]
Hi @lachezar-n! The script you posted should work, but I can reproduce the error. I assume it has the same cause as the problem described in #558. A quick fix is to insert the statement `V0=V0;` as the first line inside the for-loop. That's easily done by hand, but since this DaphneDSL script is automatically generated by DaphneLib, we shouldn't waste time on implementing a workaround. Instead, I will fix this bug (#558) in the next days, it's anyway on my agenda.

Sorry for not having answered your question regarding multi-return and creating a PR yet. We welcome PRs at any stage of the development (can also be marked as draft) and are happy to give feedback on the code. I will have a look at your solution in your recently opened PR #574 soon.

**Comment by pdamme** [comment]
@lachezar-n FYI: The bug that caused the error mentioned above (`operand #0 does not dominate this use`) (#558) should be fixed now on main.
